src/uct/tcp/tcp_ep.c

```diff
@@ -329,132 +426,200 @@ void uct_tcp_ep_set_failed(uct_tcp_ep_t *ep)
+static inline void uct_tcp_ep_ctx_move(uct_tcp_ep_ctx_t *to_ctx,
+                                       uct_tcp_ep_ctx_t *from_ctx)
 {
-    uct_tcp_ep_t *ep = NULL;
-    ucs_status_t status;
-    int fd;
+    if (!uct_tcp_ep_ctx_buf_need_progress(from_ctx)) {
+        return;
+    }
+
+    memcpy(to_ctx, from_ctx, sizeof(*to_ctx));
+    memset(from_ctx, 0, sizeof(*from_ctx));
+}
 
-    /* if EP is already allocated, dest_addr can be NULL */
-    ucs_assert((*ep_p != NULL) || (dest_addr != NULL));
+static ucs_status_t uct_tcp_ep_create_socket_and_connect(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    ucs_status_t status;
 
-    status = ucs_socket_create(AF_INET, SOCK_STREAM, &fd);
+    status = ucs_socket_create(AF_INET, SOCK_STREAM, &ep->fd);
     if (status != UCS_OK) {
-        goto out;
+        goto err;
     }
 
-    if (*ep_p == NULL) {
-        status = uct_tcp_ep_init(iface, fd, dest_addr, &ep);
-        if (status != UCS_OK) {
-            ucs_close_fd(&fd);
-            goto out;
-        }
-
-        ep->conn_sn = conn_sn;
-
-        /* EP is responsible for this socket fd from now */
-        fd = -1;
-    } else {
-        ep     = *ep_p;
-        ep->fd = fd;
+    status = uct_tcp_iface_set_sockopt(iface, ep->fd);
+    if (status != UCS_OK) {
+        goto err;
     }
 
     status = uct_tcp_cm_conn_start(ep);
     if (status != UCS_OK) {
+        goto err;
+    }
+
+out:
+    return status;
+
+err:
+    if (ep->conn_retries > 1) {
+        /* if this is not the first connection establishment retry (i.e. it
+         * is not called from uct_ep_create()/uct_ep_connect_to_ep()), set
+         * EP as failed */
         uct_tcp_ep_set_failed(ep);
-        goto out;
     }
+    goto out;
+}
+
+void uct_tcp_ep_replace_ep(uct_tcp_ep_t *to_ep, uct_tcp_ep_t *from_ep)
+{
+    uct_tcp_iface_t *iface   = ucs_derived_of(to_ep->super.super.iface,
+                                              uct_tcp_iface_t);
+    int events               = from_ep->events;
+    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
+
+    uct_tcp_ep_mod_events(from_ep, 0, from_ep->events);
+    to_ep->fd   = from_ep->fd;
+    from_ep->fd = -1;
+    uct_tcp_ep_mod_events(to_ep, events, 0);
 
-    if (*ep_p == NULL) {
-        *ep_p = ep;
+    to_ep->conn_retries++;
+
+    uct_tcp_ep_ctx_move(&to_ep->tx, &from_ep->tx);
+    uct_tcp_ep_ctx_move(&to_ep->rx, &from_ep->rx);
+
+    ucs_queue_splice(&to_ep->pending_q, &from_ep->pending_q);
+    ucs_queue_splice(&to_ep->put_comp_q, &from_ep->put_comp_q);
+
+    to_ep->flags |= from_ep->flags & (UCT_TCP_EP_FLAG_ZCOPY_TX           |
+                                      UCT_TCP_EP_FLAG_PUT_RX             |
+                                      UCT_TCP_EP_FLAG_PUT_TX_WAITING_ACK |
+                                      UCT_TCP_EP_FLAG_PUT_RX_SENDING_ACK);
+
+    if (uct_tcp_ep_ctx_buf_need_progress(&to_ep->rx)) {
```

Comment:
why would we read additional data? can we read only up to TCP connect request packet size?